// swc-minify-webpack-plugin/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  SwcMinifyWebpackPlugin: () => SwcMinifyWebpackPlugin
});
module.exports = __toCommonJS(src_exports);
var import_core = require("@swc/core");
var import_webpack = __toESM(require("webpack"));
var { RawSource, SourceMapSource } = import_webpack.default.sources;
var isJsFile = /\.[cm]?js(\?.*)?$/i;
var SwcMinifyWebpackPlugin = class {
  constructor(options = {}) {
    this.options = {
      compress: true,
      mangle: true
    };
    Object.assign(this.options, options);
  }
  apply(compiler) {
    const pluginName = this.constructor.name;
    const meta = JSON.stringify({
      name: pluginName,
      version: "2.1.3",
      options: this.options
    });
    compiler.hooks.compilation.tap(pluginName, (compilation) => {
      compilation.hooks.chunkHash.tap(pluginName, (_, hash) => hash.update(meta));
      compilation.hooks.processAssets.tapPromise(
        {
          name: pluginName,
          stage: compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
          additionalAssets: true
        },
        () => this.optimize(compiler, compilation)
      );
      compilation.hooks.statsPrinter.tap(pluginName, (stats) => {
        stats.hooks.print.for("asset.info.minimized").tap("swc-minify-webpack-plugin", (minimized, { green, formatFlag }) => {
          if (minimized) {
            if (green && formatFlag) {
              return green(formatFlag("minimized"));
            } else {
              return "minimized";
            }
          } else {
            return "";
          }
        });
      });
    });
  }
  optimize(compiler, compilation) {
    const {
      options: { devtool }
    } = compiler;
    const sourceMap = this.options.sourceMap === void 0 ? !!devtool && devtool.includes("source-map") : this.options.sourceMap;
    const assets = compilation.getAssets().filter((asset) => !asset.info.minimized && isJsFile.test(asset.name));
    return this.processAssets(assets, sourceMap, compilation);
  }
  async processAssets(assets, sourceMap, compilation) {
    await Promise.all(
      assets.map(async (asset) => {
        const { source, map } = asset.source.sourceAndMap();
        const output = await (0, import_core.minify)(Buffer.isBuffer(source) ? source.toString() : source, {
          ...this.options,
          sourceMap
        });
        let newMap;
        if (output.map) {
          newMap = JSON.parse(output.map);
          newMap.sources = [asset.name];
        }
        const newSource = sourceMap && newMap ? new SourceMapSource(output.code, asset.name, newMap, source, map, true) : new RawSource(output.code);
        const newInfo = { ...asset.info, minimized: true };
        compilation.updateAsset(asset.name, newSource, newInfo);
      })
    );
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SwcMinifyWebpackPlugin
});
